File: src/common/main/kotlin/hauveli/fishcasting/casting/recipe/brainsweep/TaggedEntityTypeIngredient.java
package hauveli.fishcasting.casting.recipe.brainsweep;

import at.petrak.hexcasting.common.recipe.ingredient.brainsweep.BrainsweepeeIngredientType;
import at.petrak.hexcasting.common.recipe.ingredient.brainsweep.EntityTypeIngredient;

import com.mojang.serialization.Codec;
import com.mojang.serialization.MapCodec;
import com.mojang.serialization.codecs.RecordCodecBuilder;
import net.minecraft.core.registries.BuiltInRegistries;
import net.minecraft.core.registries.Registries;
import net.minecraft.network.RegistryFriendlyByteBuf;
import net.minecraft.network.codec.ByteBufCodecs;
import net.minecraft.network.codec.StreamCodec;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.world.entity.Entity;
import net.minecraft.world.entity.EntityType;

import java.util.Objects;

public class TaggedEntityTypeIngredient extends EntityTypeIngredient {
    public final String tag;

    /*
    @Override
    public Entity exampleEntity(Level level) {
        return this.entityType.create(level);
    }
     */

    public TaggedEntityTypeIngredient(EntityType<?> entityType, String tag) {
        super(entityType);
        this.tag = tag;
    }

    public String getTag() {
        return tag;
    }

    @Override
    public boolean test(Entity entity, ServerLevel level) {
        // entity types are singletons
        return entity.getType() == this.entityType
                && entity.getTags().contains(this.tag);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        TaggedEntityTypeIngredient that = (TaggedEntityTypeIngredient) o;
        return Objects.equals(entityType, that.entityType);
    }

    public static class Type implements BrainsweepeeIngredientType<TaggedEntityTypeIngredient> {
        public static final MapCodec<TaggedEntityTypeIngredient> CODEC = RecordCodecBuilder.mapCodec(instance -> instance.group(
                BuiltInRegistries.ENTITY_TYPE.byNameCodec().fieldOf("entityType").forGetter(TaggedEntityTypeIngredient::getEntityType),
                Codec.STRING.fieldOf("tag").forGetter(TaggedEntityTypeIngredient::getTag)
        ).apply(instance, TaggedEntityTypeIngredient::new));
        public static final StreamCodec<RegistryFriendlyByteBuf, TaggedEntityTypeIngredient> STREAM_CODEC = StreamCodec.composite(
                ByteBufCodecs.registry(Registries.ENTITY_TYPE), TaggedEntityTypeIngredient::getEntityType,
                ByteBufCodecs.STRING_UTF8, TaggedEntityTypeIngredient::getTag,
                TaggedEntityTypeIngredient::new
        );

        @Override
        public MapCodec<TaggedEntityTypeIngredient> codec() {
            return CODEC;
        }

        @Override
        public StreamCodec<RegistryFriendlyByteBuf, TaggedEntityTypeIngredient> streamCodec() {
            return STREAM_CODEC;
        }
    }
}

File: src/common/main/kotlin/hauveli/fishcasting/casting/recipe/brainsweep/TaggedEntityTypeIngredient.kt
package hauveli.fishcasting.casting.recipe.brainsweep

import at.petrak.hexcasting.common.recipe.ingredient.brainsweep.BrainsweepeeIngredientType
import at.petrak.hexcasting.common.recipe.ingredient.brainsweep.EntityTypeIngredient
import com.mojang.serialization.Codec
import com.mojang.serialization.MapCodec
import com.mojang.serialization.codecs.RecordCodecBuilder
import net.minecraft.core.registries.BuiltInRegistries
import net.minecraft.core.registries.Registries
import net.minecraft.network.RegistryFriendlyByteBuf
import net.minecraft.network.codec.ByteBufCodecs
import net.minecraft.network.codec.StreamCodec
import net.minecraft.server.level.ServerLevel
import net.minecraft.world.entity.Entity
import net.minecraft.world.entity.EntityType
import java.util.function.BiFunction
import java.util.function.Function


class TaggedEntityTypeIngredient(
    entityType: EntityType<*>?, val tag: String?) : EntityTypeIngredient(entityType) {
    override fun test(entity: Entity, level: ServerLevel?): Boolean {
        return entity.type === this.entityType
                && entity.tags.contains(this.tag)
    }

    override fun equals(o: Any?): Boolean {
        if (this === o) return true
        if (o == null || javaClass != o.javaClass) return false
        val that = o as TaggedEntityTypeIngredient
        return entityType == that.entityType
    }

    class Type : BrainsweepeeIngredientType<TaggedEntityTypeIngredient?> {
        override fun codec(): MapCodec<TaggedEntityTypeIngredient?>? {
            return CODEC
        }

        override fun streamCodec(): StreamCodec<RegistryFriendlyByteBuf?, TaggedEntityTypeIngredient?> {
            return STREAM_CODEC
        }

        companion object {
            val CODEC: MapCodec<TaggedEntityTypeIngredient?>? =
                RecordCodecBuilder.mapCodec<TaggedEntityTypeIngredient?>(
                    Function { instance: RecordCodecBuilder.Instance<TaggedEntityTypeIngredient?>? ->
                        instance!!.group(
                            BuiltInRegistries.ENTITY_TYPE.byNameCodec().fieldOf("entityType")
                                .forGetter<TaggedEntityTypeIngredient?>(
                                    { obj: TaggedEntityTypeIngredient? -> obj!!.getEntityType() }),
                            Codec.STRING.fieldOf("tag")
                                .forGetter<TaggedEntityTypeIngredient?>({ obj: TaggedEntityTypeIngredient? -> obj!!.tag })
                        ).apply<TaggedEntityTypeIngredient?>(
                            instance,
                            { entityType: EntityType<*>?, tag: String? ->
                                TaggedEntityTypeIngredient(
                                    entityType,
                                    tag
                                )
                            })
                    })
            val STREAM_CODEC: StreamCodec<RegistryFriendlyByteBuf?, TaggedEntityTypeIngredient?> =
                StreamCodec.composite<RegistryFriendlyByteBuf?, TaggedEntityTypeIngredient?, EntityType<*>?, String?>(
                    ByteBufCodecs.registry(Registries.ENTITY_TYPE),
                    { obj: TaggedEntityTypeIngredient? -> obj!!.getEntityType() },
                    ByteBufCodecs.STRING_UTF8,
                    { obj: TaggedEntityTypeIngredient? -> obj!!.tag },
                    { entityType: EntityType<*>?, tag: String? ->
                        TaggedEntityTypeIngredient(
                            entityType,
                            tag
                        )
                    }
                )
        }
    }
}